lib/reversetunnel/agent.go

```diff
@@ -354,16 +354,24 @@ func (a *Agent) handleGlobalRequests(ctx context.Context, requestCh <-chan *ssh.
 
 			switch r.Type {
 			case versionRequest:
-				err := r.Reply(true, []byte(teleport.Version))
+				version := teleport.Version
+
+				pong, err := a.Client.Ping(ctx)
 				if err != nil {
-					log.Debugf("Failed to reply to %v request: %v.", r.Type, err)
+					a.log.WithError(err).Debugf("Failed to ping auth server.")
+				} else {
+					version = pong.ServerVersion
+				}
```

Comment:
Here we never know which version is returned, as that can be either the the proxy or the auth server version. I'm not sure if it's worth to do that, but what do you think of introducing a new "endpoint" that returns the JSON and falls back to the current one if the server doesn't implement it. Then we can have explicit proxy and auth version.

Reply:
> @rosstimothy I was using this logic in my other PR, but this PR won't make a huge difference #12040. But there is other thing. I assume that this PR will be backported to v9. The function checks if the cluster that you're connecting to is before v8. Technically we allow only one version different, so if this PR makes only to v9, then the only cluster that you can connect to older than v9 is v8. From what I understand this is not supported scenario in our case? Correct me if I'm wrong.

This PR needs to be backported as far as possible. It was actually found by a customer upgrading from v6 -> v7. I don't think any of the changes introduced here really change the logic in which clusters you can connect to. It simply changes which service in the remote cluster is used to determine its version. Using the agent version is incorrect both for determining the cache policy, and for the new use case that you added in #12040. Since the cache is generated by connecting to the auth server, and the remote watchers are all created from the cache we need to know the auth version. Things only work as is **if** the auth version and agent version are the same.

Reply:
I thought about this as well. However, I don't think there are any use cases that require knowing the agent version and the auth version. All remote cache decisions should be made based on the remote auth version

Reply:
Can we in this case return an error if we failed to get the version from auth server? The current implementation can still return the wrong version.